sklearn/tree/tests/test_tree.py

```diff
@@ -1950,6 +1975,74 @@ def test_apply_path_readonly_all_trees(name):
     check_apply_path_readonly(name)
 
 
+@pytest.mark.parametrize("criterion", ["mse", "friedman_mse", "poisson"])
+@pytest.mark.parametrize("Tree", REG_TREES.values())
+def test_balance_property(criterion, Tree):
+    # Test that sum(y_pred)=sum(y_true) on training set.
+    # This works if the mean is predicted (should even be true for each leaf).
+    # MAE predicts the median and is therefore excluded from this test.
+
+    # Choose a training set with non-negative targets (for poisson)
+    X, y = diabetes.data, diabetes.target
+    reg = Tree(criterion=criterion)
+    reg.fit(X, y)
+    assert np.sum(reg.predict(X)) == pytest.approx(np.sum(y))
+
+
+def test_poisson_zero_nodes():
+    # Test that sum(y)=0 and therefore y_pred=0 is forbidden on nodes.
+    X = [[0, 0], [0, 1], [0, 2], [0, 3],
+         [1, 0], [1, 2], [1, 2], [1, 3]]
+    y = [0, 0, 0, 0, 1, 2, 3, 4]
+    # Note that X[:, 0] == 0 is a 100% indicator for y == 0. The tree could
+    # easily learn that.
+    reg = DecisionTreeRegressor(criterion="poisson", random_state=1)
+    reg.fit(X, y)
+    assert np.all(reg.predict(X) > 0)
+
+
+def test_poisson_vs_mse():
+    # For Poisson distributed target, Poisson loss should give better results
+    # than least squares measured in Poisson deviance as metric.
+    # Compare to
+    # sklearn/ensemble/_hist_gradient_boosting/tests/test_gradient_boosting.py
+    # Note: Some fine tuning was needed to have metric_poi < metric_dummy on
+    # the test set!
```

Comment:
Can you confirm that this was an overfitting problem, i.e. that any reasonable value > 10 for `min_samples_split` would still pass the test? Otherwise there might be some underlying issue

Reply:
For `min_sample_split>=45` MSE gives better poisson deviance on the test set, see details.
The `n_test` is the most fine-tuned one. For `n_test=100`, for instance, the dummy regressor gives better poisson deviances on the test set. Given that test deviance is 4 times higher than train deviance, this is an overfitting issue. Any suggestions?

<details>

```python
import numpy as np
from sklearn.datasets import make_low_rank_matrix
from sklearn.dummy import DummyRegressor
from sklearn.metrics import mean_poisson_deviance, mean_squared_error
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeRegressor


rng = np.random.RandomState(42)
n_train, n_test, n_features = 500, 500, 10
X = make_low_rank_matrix(n_samples=n_train + n_test,
                                  n_features=n_features, random_state=rng)
# We create a log-linear Poisson model and downscale coef as it will get
# exponentiated.
coef = rng.uniform(low=-2, high=2, size=n_features) / np.max(X, axis=0)
y = rng.poisson(lam=np.exp(X @ coef))
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=n_test,
                                                    random_state=rng)

for msp in range(2, 81):
    tree_poi = DecisionTreeRegressor(criterion="poisson",
                                     min_samples_split=msp,
                                     random_state=rng)
    tree_mse = DecisionTreeRegressor(criterion="mse",
                                     min_samples_split=msp,
                                     random_state=rng)
    tree_poi.fit(X_train, y_train)
    tree_mse.fit(X_train, y_train)
    a = mean_poisson_deviance(y_test, tree_poi.predict(X_test))
    b = mean_poisson_deviance(y_test, np.clip(tree_mse.predict(X_test), 1e-15, None))
    print(f"min_sample_split={msp:2d}  Test Deviance  poisson={a:3.2f},  mse={b: 4.2f}")
```
Output (for splitting criterion poisson and mse)
```
min_sample_split= 2  Test Deviance  poisson=5.27,  mse=21.41
min_sample_split= 3  Test Deviance  poisson=5.08,  mse=21.10
min_sample_split= 4  Test Deviance  poisson=5.15,  mse=22.67
min_sample_split= 5  Test Deviance  poisson=5.43,  mse=19.81
min_sample_split= 6  Test Deviance  poisson=4.94,  mse=17.21
min_sample_split= 7  Test Deviance  poisson=4.73,  mse=16.62
min_sample_split= 8  Test Deviance  poisson=5.30,  mse=15.79
min_sample_split= 9  Test Deviance  poisson=4.89,  mse=14.02
min_sample_split=10  Test Deviance  poisson=4.85,  mse=12.14
min_sample_split=11  Test Deviance  poisson=4.89,  mse=11.24
min_sample_split=12  Test Deviance  poisson=4.77,  mse=10.68
min_sample_split=13  Test Deviance  poisson=4.70,  mse=8.57
min_sample_split=14  Test Deviance  poisson=4.72,  mse=8.50
min_sample_split=15  Test Deviance  poisson=4.66,  mse=8.53
min_sample_split=16  Test Deviance  poisson=4.62,  mse=8.45
min_sample_split=17  Test Deviance  poisson=4.52,  mse=7.42
min_sample_split=18  Test Deviance  poisson=4.43,  mse=7.38
min_sample_split=19  Test Deviance  poisson=4.40,  mse=7.38
min_sample_split=20  Test Deviance  poisson=4.28,  mse=7.42
min_sample_split=21  Test Deviance  poisson=4.27,  mse=7.18
min_sample_split=22  Test Deviance  poisson=4.28,  mse=6.99
min_sample_split=23  Test Deviance  poisson=4.32,  mse=7.05
min_sample_split=24  Test Deviance  poisson=4.32,  mse=7.01
min_sample_split=25  Test Deviance  poisson=4.32,  mse=6.95
min_sample_split=26  Test Deviance  poisson=4.32,  mse=6.99
min_sample_split=27  Test Deviance  poisson=4.31,  mse=7.01
min_sample_split=28  Test Deviance  poisson=4.31,  mse=6.98
min_sample_split=29  Test Deviance  poisson=4.31,  mse=6.94
min_sample_split=30  Test Deviance  poisson=4.31,  mse=6.94
min_sample_split=31  Test Deviance  poisson=4.31,  mse=6.96
min_sample_split=32  Test Deviance  poisson=4.31,  mse=7.00
min_sample_split=33  Test Deviance  poisson=4.31,  mse=7.00
min_sample_split=34  Test Deviance  poisson=4.31,  mse=6.96
min_sample_split=35  Test Deviance  poisson=4.31,  mse=6.96
min_sample_split=36  Test Deviance  poisson=4.29,  mse=6.90
min_sample_split=37  Test Deviance  poisson=4.29,  mse=6.90
min_sample_split=38  Test Deviance  poisson=4.29,  mse=4.53
min_sample_split=39  Test Deviance  poisson=4.28,  mse=4.47
min_sample_split=40  Test Deviance  poisson=4.28,  mse=4.47
min_sample_split=41  Test Deviance  poisson=4.29,  mse=4.47
min_sample_split=42  Test Deviance  poisson=4.28,  mse=4.47
min_sample_split=43  Test Deviance  poisson=4.28,  mse=4.44
min_sample_split=44  Test Deviance  poisson=4.29,  mse=4.87
min_sample_split=45  Test Deviance  poisson=4.28,  mse=4.00
min_sample_split=46  Test Deviance  poisson=4.29,  mse=4.03
min_sample_split=47  Test Deviance  poisson=4.28,  mse=4.03
min_sample_split=48  Test Deviance  poisson=4.29,  mse=3.97
min_sample_split=49  Test Deviance  poisson=4.29,  mse=3.99
min_sample_split=50  Test Deviance  poisson=4.29,  mse=3.99
min_sample_split=51  Test Deviance  poisson=4.27,  mse=4.01
min_sample_split=52  Test Deviance  poisson=4.28,  mse=4.01
min_sample_split=53  Test Deviance  poisson=4.27,  mse=4.01
min_sample_split=54  Test Deviance  poisson=4.27,  mse=4.01
min_sample_split=55  Test Deviance  poisson=4.27,  mse=4.01
min_sample_split=56  Test Deviance  poisson=4.28,  mse=3.99
min_sample_split=57  Test Deviance  poisson=4.28,  mse=3.99
min_sample_split=58  Test Deviance  poisson=4.28,  mse=3.99
min_sample_split=59  Test Deviance  poisson=4.28,  mse=4.02
min_sample_split=60  Test Deviance  poisson=4.27,  mse=4.07
min_sample_split=61  Test Deviance  poisson=4.27,  mse=4.07
min_sample_split=62  Test Deviance  poisson=4.28,  mse=4.06
min_sample_split=63  Test Deviance  poisson=4.27,  mse=4.06
min_sample_split=64  Test Deviance  poisson=4.28,  mse=3.91
min_sample_split=65  Test Deviance  poisson=4.28,  mse=3.91
min_sample_split=66  Test Deviance  poisson=4.27,  mse=3.91
min_sample_split=67  Test Deviance  poisson=4.27,  mse=3.91
min_sample_split=68  Test Deviance  poisson=4.27,  mse=3.91
min_sample_split=69  Test Deviance  poisson=4.27,  mse=3.91
min_sample_split=70  Test Deviance  poisson=4.27,  mse=3.91
min_sample_split=71  Test Deviance  poisson=4.27,  mse=3.92
min_sample_split=72  Test Deviance  poisson=4.26,  mse=3.92
min_sample_split=73  Test Deviance  poisson=4.26,  mse=3.92
min_sample_split=74  Test Deviance  poisson=4.26,  mse=3.92
min_sample_split=75  Test Deviance  poisson=4.26,  mse=3.92
min_sample_split=76  Test Deviance  poisson=4.26,  mse=3.92
min_sample_split=77  Test Deviance  poisson=4.26,  mse=3.92
min_sample_split=78  Test Deviance  poisson=4.26,  mse=3.92
min_sample_split=79  Test Deviance  poisson=4.26,  mse=3.92
min_sample_split=80  Test Deviance  poisson=4.26,  mse=3.92
```
</details>